### ISSUE TITLE
C/C++ Incompatibility in Bamboo

### ISSUE THREAD

**Issue by scottb5612** [SergioMartin86/bamboo6#11]
I am now able to translate cart3d without incident.
So long as I specify -fpermissive, the code  compiles without errors.

The error message is
warning: invalid conversion from ‘void*’ to ‘int*’ [-fpermissive]
and it appears that the problem has to do with assigning pointers to null

x = ((void *)0)
where the type of x is T * for some type T. 

In some places, bamboo correctly casts types as in 

a_sRequest = ((MPI_Request *)((void *)0));

**Comment by scottb5612** [comment]
The problem is due to interlanguage issues involving null pointers
The C pre-processor substitutes
(void *) 0 in place of NULL
C++ is complaining. It should be seeing nullptr.
I've been handling things like this via conditional compilation in the application code:
#ifdef __cplusplus
#endif 

The problem is occuring in a macro that frees up storage
#define FREE(p)  if (p) {free ((char *) p);p=NULL;  }

NULL needs to be set to null, which is compatible with C++

I think we could get seriously distracted with C v C++ issues.
Though I have to deal with them with Cart3d, I think we should
talk about C++ code in the release. There are too many potential problem.

**Comment by scottb5612** [comment]
I've created a new test that illustrate the probelm.
I simply added it to the end of the argv test
I now call it t2 and you can find it here:
https://www.dropbox.com/s/myuvom4oiq3qym0/t2.c?dl=0

Scott

**Comment by scottb5612** [comment]
After hand editing (void *) 0 to 0
I was left with 4 more invalid comparison errors
I've tracked these down. These errors arise in Bamboo's translation
of Reduce.
I looked at Bamboo collective library for reduce.
The problem occurs at line 18:

 void* tempbuf= sendbuf; //(void*)malloc(count*getsizeof(datatype));

C++ won't tolerate this, but C will.
That brings us to the next issue: all the libraries are written in C!
They should be in  C++.
